airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionHandler.kt

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+package io.airbyte.cdk.integrations.util
+
+import io.airbyte.cdk.integrations.base.AirbyteTraceMessageUtility
+import io.airbyte.cdk.integrations.base.Command
+import io.airbyte.cdk.integrations.base.errors.messages.ErrorMessage
+import io.airbyte.commons.exceptions.ConfigErrorException
+import io.airbyte.commons.exceptions.ConnectionErrorException
+import io.airbyte.commons.exceptions.TransientErrorException
+import io.airbyte.protocol.models.v0.AirbyteConnectionStatus
+import io.airbyte.protocol.models.v0.AirbyteMessage
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.util.function.Consumer
+import kotlin.system.exitProcess
+import org.jetbrains.annotations.VisibleForTesting
+
+private val LOGGER = KotlinLogging.logger {}
+
+const val COMMON_EXCEPTION_MESSAGE_TEMPLATE: String =
+    "Could not connect with provided configuration. Error: %s"
+
+const val DATABASE_CONNECTION_ERROR: String =
+    "Encountered an error while connecting to the database error"
+
+enum class FailureType {
+    CONFIG,
+    TRANSIENT
+}
+
+data class ConnectorErrorProfile(
+    var errorClass: String,
+    var regexMatchingPattern: String,
+    var failureType: FailureType,
+    var externalMessage: String,
+    var sampleInternalMessage: String,
+    var referenceLink: List<String>
+)
+
+/**
+ * This abstract class defines interfaces that will be implemented by individual connectors for
+ * translating internal exception error messages to external user-friendly error messages.
+ */
+open class ConnectorExceptionHandler {
```

Comment:
wouldn't defining this as an interface make more sense?

Reply:
i'm still not an expert of kotlin, so my original thinking is, we couldn't use abstract class here, as I will need to have a default implementation of this class if a connector doesn't extends this class. So that makes choose to use `open class` here. 

Does interface allow me to do the same in kotlin?

Reply:
Pretty much, yes. Kotlin interfaces are like java interfaces but with more bells and whistles. Also, when it comes to interface implementation, Kotlin has useful tricks such as [delegation](https://kotlinlang.org/docs/delegation.html) which might make it viable for you to expose an abstract interface and a concrete closed class implementation in the CDK. 

I don't have strong opinions on what's right, just that having an `open class` here is a code smell.